### ISSUE TITLE
 error 'no applicable method for 'RMSEP' applied to an object of class "list" '

### ISSUE THREAD

**Issue by tkyzar** [bhmevik/pls#9]
I am pretty new to R and statistical machine learning so I apologize if my question is really basic.  I am trying to run 5-fold cross validation with the plsr function so that I can get a training and a test RMSEP, and then use my predict function to predict on another set of similar data.

k=5
set.seed(1)
for(i in 1:k){
  fit.pls=plsr(NO23F_2[k!=i]~., data=swmplu.trim[k!=i,])
  pls.RMSEP = RMSEP(fit.pls[k!=i])
  min_comp = which.min(pls.RMSEP$val)
  prediction=predict(fit.pls, newdata=swmplu.trim[k == i,])
  SE[[i]]=(swmplu.trim[k == i,"NO23F_2"] - prediction)^2
}

I'm getting an error 'no applicable method for 'RMSEP' applied to an object of class "list" '

I know I could use 'segments' instead of doing the loop but I want to get the train and test error rates to compare to other methods.  Can you please tell me what my error is and how I can fix it?

Thank you!

**Comment by bhmevik** [comment]
The problem causing the error is `RMSEP(fit.pls[k!=i])`. The `fit.pls` created in the line above is a fitted model on the subset of `swmplu.trim` where `k != i`, and it does not make sense to try and subset it further.  Thus `RMSEP(fit.pls)` is enough.
There is another problem, though: When you call `plsr`, you subset the data set (`data=swmplu.trim[k!=i,]`).  Within this subset of the data set, there is no reason to subset the response variable; it will only create problems later. So `fit.pls <- plsr(NO23F_2 ~ ., data = swmplu.trim[k!=i, ])` is better.
Finally, I think you have made a logical error: `k` is always 5, so your test `k != i` will always be `5 != i`, thus in the first four iterations, the test will be `FALSE`, so `swmplu.trim[k!=i, ]` will select a 0-row data frame, while in the fifth iteration, the test will be `TRUE, so now `swmplu.trim[k!=i, ]` will select the whole data set. If you want to do a "manual" cross-validation, I suggest you take a look at the section "9.4 Calling fit functions directly" of the `vignette("pls-manual")` for an example (it uses the fit functions directly, but you can use the formula variant if you want; it is just a bit slower).

**Comment by tkyzar** [comment]
Thank you for responding and providing so much explanation, this is very helpful.

I copied the code in 9.4 and put in my own table references, and now I’m getting a different error

X <- swmplu.trim
Y <- swmplu.trim$NO23F_2
ncomp <- 5
cvPreds <- matrix(nrow = nrow(X), ncol = ncomp)
for (i in 1:nrow(X)) {
  fit <- simpls.fit(X[-i,], Y[-i], ncomp = ncomp, stripped = TRUE)
  cvPreds[i,] <- (X[i,] - fit$Xmeans) %*% drop(fit$coefficients) + fit$Ymeans
}

Error in `dimnames<-.data.frame`(`*tmp*`, value = dimnames(Y) <- NULL) :
  invalid 'dimnames' given for data frame

I’m guessing it’s because I am not identifying a specific ‘X’ variable.  I want to regress the Y on all the other columns.  Is my guess correct?  If so, how can I resolve this error?

Thank you!

Tricia Kyzar
Ph: 352-392-7260
Email: tkyzar@ufl.edu<mailto:tkyzar@ufl.edu>

**Comment by tkyzar** [comment]
Thank you, bhmevik

Your explanation about the unnecessary additional subset of Y makes sense and I've taken that out.  

Also, your explanation about k always being equal to 5 causing the plsr to fail on the 5th iteration makes sense.  So I've been experimenting with the swmplu.trim[k!=i, ] and have discovered that 
     data=swmplu.trim[k!=(i-1)] 
will run 5 iterations successfully, but it appears to me the iterated 'i' is not carrying through the rest of the logic because it returns the same values through all 5 iterations.  So when I run: 

     SE=list()
     k=5
     set.seed(1)
     for(i in 1:k){
       print("i: ");print(i)
  
       fit.pls=plsr(NO23F_2~., data=swmplu.trim[k!=(i-1)])
       print("fit.pls: ");print(fit.pls$Ymeans[k!=(i-1)])
  
       pls.RMSEP = RMSEP(fit.pls)
       min_comp = which.min(pls.RMSEP$val)
       print("min_comp: "); print(min_comp)
  
       prediction=predict(fit.pls, newdata=swmplu.trim[k == i,])
       SE[[i]]=(swmplu.trim[k == i,"NO23F_2"] - prediction)^2
     }

I get:
     [1] "i: "
     [1] 1
     [1] "fit.pls: "
     [1] 0.01775782
     [1] "min_comp: "
     [1] 20
     [1] "i: "
     [1] 2
     [1] "fit.pls: "
     [1] 0.01775782
     [1] "min_comp: "
     [1] 20
     [1] "i: "
     [1] 3
     [1] "fit.pls: "
     [1] 0.01775782
     [1] "min_comp: "
     [1] 20
     [1] "i: "
     [1] 4
     [1] "fit.pls: "
     [1] 0.01775782
     [1] "min_comp: "
     [1] 20
     [1] "i: "
     [1] 5
     [1] "fit.pls: "
     [1] 0.01775782
     [1] "min_comp: "
     [1] 20

Can you please tell me what is my error here and how I can fix it?

Thank you!

**Comment by bhmevik** [comment]
Well, `k` is still constantly 5, so the test `k!=(i-1)` will always be true, thus the five fits will all run on the whole data set (that is why you get the same result every time). Did you take a look at how the example in the vignette does the subsetting?

**Comment by bhmevik** [comment]
Yes, it looks like it should be right, with one exception: Usually with k-fold CV one tries to select the folds so they have the same number of elements.  With `folds=sample(1:5, nrow(swmplu.trim), replace=TRUE)` you are not guaranteed this.  If you try `table(folds)` you will probably see that the five numbers occur a different number of times.
Instead, you can use `folds <- sample(rep(1:5, length = nrow(swmplu.trim)))`. The `rep()` will repeat `1:5` until it gets `nrow(swmplu.trim)` elements, and the `sample()` will simply reorder the elements randomly.

**Comment by tkyzar** [comment]
Dr. Mevik,

I have my plsr with the sample(rep(...)) code as indicated and I'm curious.  My "min_comp = which.min(pls.RMSEP$val)" returns that 20 components has the lowest pls.RMSEP$val, but I see that 19 components has the same value (unless it's beyond what is currently displayed for significant digits).  So my question is two fold: 20 components seems like a pretty high component count, am I misunderstanding what is being returned to me?  And why would it not return 19 for the which.min instead of 20?

plsSE=list()
set.seed(1)
folds = sample(rep(1:5, length = nrow(swmplu.trim)))
for(i in 1:5){
  print(paste0("i: ",i))
  
  fit.pls=plsr(NO23F_2~., data=swmplu.trim[folds!=i,])
  print(paste0("fit.pls: ",fit.pls$Ymeans))
  
  pls.RMSEP = RMSEP(fit.pls)
  min_comp = which.min(pls.RMSEP$val)
  print(paste0("min_comp: ", min_comp = which.min(pls.RMSEP$val)))
  
  prediction=predict(fit.pls, newdata=swmplu.trim[folds==i,], min_comp)
  plsSE[[i]]=mean(swmplu.trim[i,"NO23F_2"] - prediction)^2
}

Thank you, again for all your time and patience explaining all this to me!

**Comment by bhmevik** [comment]
Unless the data set is huge, 20 components is a lot, yes. The reson which.min returned 20 instead of 19 has to be because the value for 20 is less than for 19 (or perhaps equal, but then I _think_ it would return 19 -- check with help("which.min")!). But choosing the global minimum is rarely the best strategy - you would do better plotting the RMSEPs and choosing a value near the place where the curve starts to flatten out, i.e., where adding more components doesn't give you much improvement. How to choose the number of components has been discussed in numerous introductions to PLSR/PCR, and in many papers - I would suggest finding an introduction. There are also some automated procedures; the pls package includes one.

**Comment by khliland** [comment]
A technical note here is that RMSEP$val includes the intercept, so your min_comp is offset by 1.

**Comment by tkyzar** [comment]
Thank you, again for your responses.  khliland - your note is very interesting.  I noticed when I used the pcr function (all the same setup as my pls, just using pcr instead of plsr), my min_comp returned 44 - which is not logical because I have a total of 44 variables in my table, which includes my response variable.  Your note explains the min_comps for both of my models.

I went back to the pls introduction by Dr. Mevik and added the 'selectNcomp' function, but I get an error message ("No cross-validation data available in model") which I assume is because I have plsr function in a loop instead of in it's original form.    I'll keep looking for other options.

Thank you!
